# Incremental Capstone — Session 2

## Data Processing and Statistical Analysis

### Objective

The goal of Session 2 is to continue processing the prepared NSMES1988 healthcare dataset created during Session 1.

In this notebook, we will:

- Import the required Python libraries
- Load `NSMES1988new.csv`
- Compare memory usage with Session 1
- Convert `age` and `income` to their intended units
- Perform basic statistical analysis
- Use `describe()` to summarize the numerical data
- Identify variables that are not appropriate for numerical statistical analysis
- Review possible data type changes
- Export an updated dataset for later capstone sessions
- Document findings in Markdown cells

## 1. Import Required Libraries

Pandas and NumPy will be imported for data processing and statistical analysis. Pandas is used to work with DataFrames, while NumPy provides numerical operations that may be useful during the analysis.

In [1]:
# Import Pandas and use the standard alias "pd".
# Pandas provides tools for loading, organizing, cleaning,
# and analyzing tabular datasets.
import pandas as pd

# Import NumPy and use the standard alias "np".
# NumPy provides numerical and statistical operations.
import numpy as np

## 2. Import the Prepared Dataset

The `NSMES1988new.csv` file created during Session 1 will be imported into a Pandas DataFrame. This dataset contains the prepared healthcare data that will be used for the data processing and statistical analysis performed during Session 2.

The first five rows will be displayed after importing the file to verify that the dataset loaded successfully.

In [2]:
# Read the NSMES1988new.csv file created during Session 1
# and store the data in a Pandas DataFrame named df.
df = pd.read_csv("../data/NSMES1988new.csv")

# Display the first five rows of the DataFrame
# to verify that the file loaded correctly.
df.head()

,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,5,0,0,0,0,1,average,2,normal,other,6.9,male,yes,6,2.8810,yes,yes,no
1,1,0,2,0,2,0,average,2,normal,other,7.4,female,yes,10,2.7478,no,yes,no
2,13,0,0,0,3,3,poor,4,limited,other,6.6,female,no,10,0.6532,no,no,yes
3,16,0,5,0,1,1,poor,2,limited,other,7.6,male,yes,3,0.6588,no,yes,no
4,3,0,0,0,0,0,average,2,limited,other,7.9,female,yes,6,0.6588,no,yes,no


## 3. Memory Analysis of the New DataFrame

The memory usage and data types of the newly imported `NSMES1988new.csv` dataset will be examined and compared with the results from Session 1.

Deep memory inspection will be used to provide a more accurate measurement of memory consumption, particularly for columns containing string values.

In [3]:
# Display structural information about the newly loaded DataFrame.
# memory_usage="deep" provides a more accurate calculation of memory
# usage by including the memory consumed by string values.
df.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 18 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   visits     4406 non-null   int64  
 1   nvisits    4406 non-null   int64  
 2   ovisits    4406 non-null   int64  
 3   novisits   4406 non-null   int64  
 4   emergency  4406 non-null   int64  
 5   hospital   4406 non-null   int64  
 6   health     4406 non-null   str    
 7   chronic    4406 non-null   int64  
 8   adl        4406 non-null   str    
 9   region     4406 non-null   str    
 10  age        4406 non-null   float64
 11  gender     4406 non-null   str    
 12  married    4406 non-null   str    
 13  school     4406 non-null   int64  
 14  income     4406 non-null   float64
 15  employed   4406 non-null   str    
 16  insurance  4406 non-null   str    
 17  medicaid   4406 non-null   str    
dtypes: float64(2), int64(8), str(8)
memory usage: 2.4 MB


### Memory Comparison with Session 1

The newly imported `NSMES1988new.csv` DataFrame contains 4,406 observations and 18 columns and uses approximately 2.4 MB of memory.

At the end of Session 1, the optimized in-memory DataFrame used approximately 173.6 KB after categorical variables were converted to the Pandas `category` data type and smaller whole-number variables were converted to `int16`.

After the optimized DataFrame was exported to CSV and reloaded during Session 2, Pandas inferred the categorical variables as strings and the integer variables as `int64`. As a result, memory usage increased from approximately 173.6 KB to 2.4 MB.

This demonstrates that CSV files preserve the data values but do not preserve Pandas-specific data type settings such as `category` and `int16`. These data types can be reapplied after importing the CSV if memory optimization is required.

## 4. Transform Age and Income

According to the dataset description, the `age` variable is stored as age in years divided by 10, while the `income` variable represents family income in units of $10,000.

To convert these variables into their intended units, `age` will be multiplied by 10 and `income` will be multiplied by 10,000.

The original values will first be displayed so that the results can be compared after the transformation.

In [4]:
# Display the first five age and income values before transformation.
# This allows us to compare the original scaled values with the
# converted values after multiplication.
df[["age", "income"]].head()

,age,income
0,6.9,2.8810
1,7.4,2.7478
2,6.6,0.6532
3,7.6,0.6588
4,7.9,0.6588


In [5]:
# Convert age from its scaled value to actual age in years.
# For example, an age value of 6.9 becomes 69 years.
df["age"] = df["age"] * 10

# Convert income from units of $10,000 to actual dollar values.
# For example, an income value of 2.8810 becomes $28,810.
df["income"] = df["income"] * 10000

In [6]:
# Display the first five age and income values after transformation.
# This verifies that age is now represented in years
# and income is now represented in actual dollar values.
df[["age", "income"]].head()

,age,income
0,69.0,28810.0
1,74.0,27478.0
2,66.0,6532.0
3,76.0,6588.0
4,79.0,6588.0


### Age and Income Transformation Results

The `age` and `income` variables were converted from their original scaled units into more interpretable values.

The `age` variable was multiplied by 10 so that it now represents actual age in years. For example, an original value of 6.9 was converted to 69 years.

The `income` variable was multiplied by 10,000 so that it now represents family income in actual U.S. dollars. For example, an original value of 2.8810 was converted to $28,810.

The transformed values were reviewed to confirm that the calculations were applied successfully.

## 5. Basic Statistical Analysis

Basic statistical measures will be calculated for the numerical variables in the dataset. These measures will help summarize the center, spread, and range of the numerical data.

The analysis will include the count, mean, median, standard deviation, minimum, and maximum values for each numerical column.

In [7]:
# Select only the numerical columns from the DataFrame.
# This excludes string variables such as health, gender, and region.
numeric_df = df.select_dtypes(include="number")

# Calculate several basic statistical measures for each numerical column.
basic_stats = numeric_df.agg([
    "count",
    "mean",
    "median",
    "std",
    "min",
    "max"
])

# Display the statistical results.
basic_stats

,visits,nvisits,ovisits,novisits,emergency,hospital,chronic,age,school,income
count,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000
mean,5.774399,1.618021,0.750794,0.536087,0.263504,0.295960,1.541988,74.024058,10.290286,25271.320468
median,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,73.000000,11.000000,16981.500000
std,6.759225,5.317056,3.652759,3.879506,0.703659,0.746398,1.349632,6.334050,3.738736,29246.476178
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,66.000000,0.000000,-10125.000000
max,89.000000,104.000000,141.000000,155.000000,12.000000,8.000000,8.000000,109.000000,18.000000,548351.000000


### Basic Statistical Analysis Findings

The numerical variables contain 4,406 observations each, indicating that there are no missing values among the numerical columns.

The average age of individuals in the dataset is approximately 74.02 years, with a median age of 73 years. Ages range from 66 to 109 years, indicating that the dataset represents an older population.

Average family income is approximately $25,271, while the median income is approximately $16,982. Income ranges from -$10,125 to $548,351. The substantially higher mean compared with the median, along with the large maximum value, suggests that the income distribution may be right-skewed. The presence of negative income values should also be considered during further analysis.

Healthcare utilization varies considerably across individuals. Physician office visits (`visits`) average approximately 5.77 visits, with a median of 4 and a maximum of 89. Several other healthcare utilization variables have median values of zero but relatively large maximum values, suggesting that many individuals did not use those services while a smaller number had much higher utilization.

The number of chronic conditions ranges from 0 to 8, with an average of approximately 1.54 conditions. Years of education (`school`) range from 0 to 18, with an average of approximately 10.29 years.

## 6. Save the Updated Dataset

After converting the `age` and `income` variables to their intended units and completing the basic statistical analysis, the updated DataFrame will be saved as `NSMES1988updated.csv`.

This file will preserve the updated data for possible use during later capstone sessions.

In [11]:
# Save the current DataFrame as NSMES1988updated.csv.
# The DataFrame contains the transformed age and income values.
# index=False prevents the Pandas row index from being
# written as an additional column in the CSV file.
df.to_csv("../data/NSMES1988updated.csv", index=False)

In [12]:
# Read the newly saved CSV file into a temporary DataFrame.
# This verifies that NSMES1988updated.csv was created correctly.
df_updated = pd.read_csv("../data/NSMES1988updated.csv")

# Display the first five rows of the saved dataset.
df_updated.head()

,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,5,0,0,0,0,1,average,2,normal,other,69.0,male,yes,6,28810.0,yes,yes,no
1,1,0,2,0,2,0,average,2,normal,other,74.0,female,yes,10,27478.0,no,yes,no
2,13,0,0,0,3,3,poor,4,limited,other,66.0,female,no,10,6532.0,no,no,yes
3,16,0,5,0,1,1,poor,2,limited,other,76.0,male,yes,3,6588.0,no,yes,no
4,3,0,0,0,0,0,average,2,limited,other,79.0,female,yes,6,6588.0,no,yes,no


## 7. Statistical Summary Using `describe()`

The Pandas `describe()` method will be used to generate a standard statistical summary of the numerical variables.

The results from `describe()` will then be compared with the basic statistical analysis performed in the previous section.

In [13]:
# Generate descriptive statistics for the numerical columns.
# describe() automatically calculates count, mean, standard deviation,
# minimum, quartiles, and maximum values.
describe_stats = numeric_df.describe()

# Display the descriptive statistics.
describe_stats

,visits,nvisits,ovisits,novisits,emergency,hospital,chronic,age,school,income
count,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000
mean,5.774399,1.618021,0.750794,0.536087,0.263504,0.295960,1.541988,74.024058,10.290286,25271.320468
std,6.759225,5.317056,3.652759,3.879506,0.703659,0.746398,1.349632,6.334050,3.738736,29246.476178
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,66.000000,0.000000,-10125.000000
25%,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,69.000000,8.000000,9121.500000
50%,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,73.000000,11.000000,16981.500000
75%,8.000000,1.000000,0.000000,0.000000,0.000000,0.000000,2.000000,78.000000,12.000000,31728.500000
max,89.000000,104.000000,141.000000,155.000000,12.000000,8.000000,8.000000,109.000000,18.000000,548351.000000


### Comparison of Basic Statistics and `describe()`

The results produced by `describe()` are consistent with the basic statistical analysis performed previously. The count, mean, standard deviation, minimum, and maximum values are the same in both analyses. The median calculated previously corresponds to the `50%` value reported by `describe()`.

The primary difference is that `describe()` also provides the 25th and 75th percentiles. These quartiles provide additional information about the distribution of the numerical variables.

For age, the 25th percentile is 69 years, the median is 73 years, and the 75th percentile is 78 years. Therefore, the middle 50% of the observed ages fall between approximately 69 and 78 years.

For income, the 25th percentile is approximately $9,121.50, the median is $16,981.50, and the 75th percentile is $31,728.50. The maximum income of $548,351 is substantially greater than the 75th percentile and median, supporting the earlier observation that the income distribution may be right-skewed.

Several healthcare utilization variables, including outpatient, non-physician outpatient, emergency, and hospital visits, have a 75th percentile of zero. This indicates that at least 75% of observations report no utilization for these particular services, while a smaller portion of individuals accounts for the positive values.

Overall, `describe()` confirms the results of the basic statistical analysis while providing additional information about the distribution of the data through quartiles.

## 8. Review Columns and Data Types

The data types of all variables will be reviewed to determine which columns are appropriate for numerical statistical analysis and which variables represent categorical information.

The review will also identify possible data type changes that may better represent the meaning of each variable.

In [14]:
# Display the data type assigned to each column in the DataFrame.
# This is an inspection step and does not modify the dataset.
df.dtypes

visits         int64
nvisits        int64
ovisits        int64
novisits       int64
emergency      int64
hospital       int64
health           str
chronic        int64
adl              str
region           str
age          float64
gender           str
married          str
school         int64
income       float64
employed         str
insurance        str
medicaid         str
dtype: object

In [15]:
# Create a list containing the columns that represent categories.
categorical_columns = [
    "health",
    "adl",
    "region",
    "gender",
    "married",
    "employed",
    "insurance",
    "medicaid"
]

# Count the number of unique values in each categorical column.
# This helps confirm that these columns contain a limited
# number of repeated categories rather than continuous measurements.
df[categorical_columns].nunique()

health       3
adl          2
region       4
gender       2
married      2
employed     2
insurance    2
medicaid     2
dtype: int64

### Variables Not Appropriate for Numerical Statistical Analysis

The variables `health`, `adl`, `region`, `gender`, `married`, `employed`, `insurance`, and `medicaid` represent categorical information rather than numerical measurements. Therefore, numerical statistics such as mean, median, and standard deviation are not appropriate for these variables.

Inspection of the number of unique values confirms that these variables contain a small number of repeated categories. `health` contains 3 unique values, `region` contains 4, and each of the remaining categorical variables contains only 2 unique values.

Although Pandas currently represents these variables as strings, the `category` data type would more accurately represent their analytical meaning and may also reduce memory usage.

The variables `visits`, `nvisits`, `ovisits`, `novisits`, `emergency`, `hospital`, `chronic`, `age`, `school`, and `income` represent numerical counts or measurements and are appropriate for numerical statistical analysis.

## Session 2 — Data Processing and Statistical Analysis Report

During Session 2, the prepared `NSMES1988new.csv` dataset from Session 1 was imported for additional processing and statistical analysis. The dataset contained 4,406 observations and 18 variables.

### Memory Analysis

Deep memory analysis showed that the newly imported CSV used approximately 2.4 MB of memory. This was higher than the approximately 173.6 KB used by the optimized in-memory DataFrame at the end of Session 1.

The increase occurred because CSV files preserve data values but do not preserve Pandas-specific data types. When the CSV was reloaded, the previously optimized `category` variables were interpreted as strings and the `int16` variables were interpreted as `int64`.

### Age and Income Transformation

The `age` variable was multiplied by 10 to convert the values into actual years. After transformation, the observed ages ranged from 66 to 109 years, with an average age of approximately 74.02 years.

The `income` variable was multiplied by 10,000 to convert the values into actual U.S. dollar amounts. Average family income was approximately $25,271, while median income was approximately $16,982. Income ranged from -$10,125 to $548,351.

### Statistical Analysis

Basic statistical analysis was performed on the numerical variables using count, mean, median, standard deviation, minimum, and maximum values.

Healthcare utilization varied substantially among individuals. Physician office visits (`visits`) averaged approximately 5.77, with a median of 4 and a maximum of 89. Several other healthcare utilization variables had median values of zero, suggesting that many individuals did not use those services while a smaller number had considerably greater utilization.

The number of chronic conditions ranged from 0 to 8, with an average of approximately 1.54. Years of education (`school`) ranged from 0 to 18, with an average of approximately 10.29 years.

The `describe()` method confirmed the results of the basic statistical analysis and additionally provided the 25th and 75th percentiles. The income statistics suggest a right-skewed distribution because the mean income is greater than the median and the maximum income is substantially greater than the 75th percentile.

### Data Type Review

The variables `health`, `adl`, `region`, `gender`, `married`, `employed`, `insurance`, and `medicaid` represent categorical information and are not appropriate for numerical statistics such as mean, median, or standard deviation.

These variables contain only a small number of unique values and could appropriately be represented using the Pandas `category` data type during analysis. The remaining numerical variables represent counts or measurements and are appropriate for numerical statistical analysis.

### Updated Dataset

After completing the required transformations, the updated DataFrame was exported as `NSMES1988updated.csv` without the Pandas index. This file contains the transformed age and income values and is available for use during subsequent capstone sessions.